text/3845-repr-ordered-fields.md

```diff
@@ -0,0 +1,381 @@
+
+Introduce two new warnings
+1. An edition migration warning, when updating to the next edition, that the meaning of `repr(C)` is changing
+2. A warn-by-default lint when `repr(ordered_fields)` is used on enums without the tag type specified. Since this is likely not what the user wanted
+3. A warn-by-default lint when `repr(C)` is used, and there are no `extern` blocks or functions in the crate (on all editions)
+
+# Motivation
+[motivation]: #motivation
+
+Currently `repr(C)` serves two roles
+1. Provide a consistent, cross-platform, predictable layout for a given type
+2. Match the target C compiler's struct/union layout algorithm and ABI
+
+But in some cases, these two cases are in tension due to platform weirdness (even on major platforms like Windows MSVC)
+* https://github.com/rust-lang/unsafe-code-guidelines/issues/521
+* https://github.com/rust-lang/rust/issues/81996
+
+Code in case 1 generally falls into one of these buckets:
+* rely on the exact layout being consistent across platforms
+	* for example, zero-copy deserialization (see [rkyv](https://crates.io/crates/rkyv))
+* manually calculating the offsets of fields
+	* This is common in code written before the stabilization of `offset_of` (currently only stabilized for `struct`)/`&raw const`/`&raw mut`
+	* But sometimes this is still required if you are manually doing type-erasure and handling DSTs (for example, implementing [`erasable::Erasable`](https://docs.rs/erasable/1.3.0/erasable/trait.Erasable.html))
+* manually calculating the layout for a DST, to prepare an allocation (see [slice-dst](https://crates.io/crates/slice-dst), specifically [here](https://github.com/CAD97/pointer-utils/blob/0fe399f8f7e519959224069360f3900189086683/crates/slice-dst/src/lib.rs#L162-L163))
+* match layouts of two different types (or even, two different monomorphizations of the same generic type)
+	* see [here](https://github.com/rust-lang/rust/pull/68099), where in `alloc` this is done for `Rc` and `Arc` to give a consistent layout for all `T`
+
+So, providing any fix for case 2 would subtly break any users of case 1. This breakage cannot be checked easily since it relies on unsafe code making assumptions about data layouts, which makes this difficult to fix within a single edition.
+
+As an example of this tension: on Windows MSVC, `repr(C)` doesn't always match what MSVC does for  ZST structs (see this [issue](https://github.com/rust-lang/rust/issues/81996) for more details)
+
+```rust
+// should have size 8, but has size 0
+#[repr(C)]
+struct SomeFFI([i64; 0]);
+```
+
+Of course, making `SomeFFI` size 8 doesn't work for anyone using `repr(C)` for case 1. They want it to be size 0 (as it currently is). 
```

Comment:
```suggestion
Of course, making `SomeFFI` size 8 doesn't work for anyone using `repr(C)` for case 1. They want it to be size 0 (as it currently is). 

Fixing the layout of this struct will also most likely let us remove a long-standing hack from our implementation of the MSVC ABI:
unlike all other ABIs, we do *not* entirely skip ZST with that ABI but instead mark them to be passed by-ptr.
This is needed to correctly pass types like `SomeFFI`.
If we fix our layout computation to match that of MSVC, we no longer need this special case in the ABI logic.
```

Reply:
That issue should presumably be addressed in this RFC, no? `repr(C)` 1-ZSTs need to be passed by pointer, but does that apply to `repr(ordered_fields)` as well?

Reply:
The main problem is that we have already specified that all 1-ZST have the same ABI. We need to figure out how to clean that up but I don't think this RFC needs to do that.

Reply:
This RFC currently specifies that new `repr(C)` matches C. That requires breaking the “all 1-ZST have the same ABI” rule. So the RFC either needs to specify that it’s actually “`repr(C)` matches C except for 1-ZSTs”, or it needs to specify that we are breaking the rule. Can’t have it both ways

Reply:
@RustyYato see the issue description of https://github.com/rust-lang/unsafe-code-guidelines/issues/552.

@Jules-Bertholet this RFC doesn't take a clear stance on "types that are not standard C but are allowed by some C compilers for that target", does it?

Reply:
>  Though ABI is a cross-edition concern so it'd only help marginally to do this on new editions only.

I assume it would be based on the edition where the `repr(C)` 1-ZST is defined? I.e., same as everything else in this RFC. Note that, if you have:

```rust
// on new edition with new `repr(C)`
#[repr(C)]
struct NoFields {}

#[repr(transparent)]
struct Foo(u32, NoFields);
```

`NoFields` won’t be a ZST *at all* on `windows-msvc`, so it’s not really much more breaking to say `Foo` doesn’t work on `windows-gnu` either.

Reply:
> Bumping the edition of the crate that has the empty C struct could lead to downstream build failures.

Well, yes, the whole point of this RFC is that changing the meaning of `repr(C)` is a breaking change that must be done over an edition.